Mathlib/Algebra/MvPolynomial/SchwartzZippel.lean

```diff
@@ -0,0 +1,123 @@
+/-
+Copyright (c) 2023 Bolton Bailey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bolton Bailey, Yaël Dillies
+-/
+import Mathlib.Algebra.MvPolynomial.Equiv
+import Mathlib.Algebra.Polynomial.Roots
+import Mathlib.Data.Fin.Tuple.Finset
+
+/-!
+# The Schwartz-Zippel lemma
+
+This file contains a proof of the
+[Schwartz-Zippel](https://en.wikipedia.org/wiki/Schwartz%E2%80%93Zippel_lemma) lemma.
+
+This lemma tells us that the probability that a nonzero multivariable polynomial over an integral
+domain evaluates to zero at a random point is bounded by the degree of the polynomial over the size
+of the field, or more generally, that a nonzero multivariable polynomial over any integral domain
+has a low probability of being zero when evaluated at points drawn at random from some finite subset
+of the field. This lemma is useful as a probabilistic polynomial identity test.
+
+## TODO
+
+* Generalize to noncommutative rings? Is that even possible?
+* Generalize to subset of the ring being different for each variable.
+  * What does the theorem say in this case?
```

Comment:
It means bounding above the set of solutions in a product type. A particular case would imply that if $$f$$ is a polynomial in $$T_1,\dots,T_n$$, $$A_1,\dots,A_n$$ are finite sets of cardinalities $$|A_i|>\deg_{T_i}(f)$$ for all $$i$$, then $$f$$ doesn't vanish identically on $$A_1\times\dots\times A_n$$. This is proved in a PR of mine, #16177, as one step of Alon's combinatorial Nullstellensatz.

Reply:
Are you sure it's a "particular case"? When applying Schwartz-Zippel, I find myself having to prove that $\sum_i \frac{\deg_{T_i}(f)}{|A_i|} < 1$, which doesn't seem true given the mere $\deg_{T_i}(f) < |A_i|$ assumption.

Reply:
Here is my code, if you want to complete it (with all the `open` from this PR)
```
/-- A corollary of the **Schwartz-Zippel** lemma. -/
lemma exists_mem_piFinset_eval_ne_zero_of_degreeOf_lt_card {n} {p : MvPolynomial (Fin n) R}
    (hp : p ≠ 0) {S : Fin n → Finset R} (hS : ∀ i, (S i).Nonempty) (hpS : p.degreeOf < (#S ·)) :
    ∃ f ∈ S ^^ n, eval f p ≠ 0 := by
  by_contra! h
  refine lt_irrefl (1 : ℚ≥0) ?_
  calc
    (1 : ℚ≥0) = #{f ∈ S ^^ n | eval f p = 0} / ∏ i, (#S i : ℚ≥0) := by
      rw [filter_eq_self.2 h, card_piFinset, Nat.cast_prod, div_self]
      exact prod_ne_zero_iff.2 fun i _ ↦ mod_cast (hS _).card_ne_zero
    _ ≤ ∑ i, (p.degreeOf i / #S i : ℚ≥0) := schwartz_zippel hp _
    _ < 1 := sorry -- Are you sure this is true?
```

Reply:
No I'm not sure at all. It was rather a guess based on my understanding of what something such as Schwartz-Zippel says (or should say) than an affirmation.

Reply:
The following paper indicates results that mathlib could host.

https://arxiv.org/abs/2305.10900

Reply:
Ahahah. This is the paper that Bhavik and I were looking at to try beating your combinatorial nullstellensatz PR 🙈

Reply:
There's nothing to beat. Let's be adults and do it.

Reply:
All of this is basic and easy math. The only problem is that it is often written like sh*t. (This paper is nicer, and the one by Pete Clark is cleaner too.)

Reply:
> There's nothing to beat. Let's be adults and do it.

To be clear, we were working on the generalisation just to spite you, but rather because we started working on it long before you opened your PR, and we thought it would be a good way to just keeping working on it

> The only problem is that it is often written like sh*t.

Yes, agree 😓

Reply:
Yes, I'm not sure where "beat" comes from here - I needed a generalisation of the combinatorial nullstellensatz (one of the ones in that paper) for an application, and started working on it on my own just before your PR. But yours develops a useful amount of theory, and I see no need to block it based off an incomplete generalisation which uses different machinery.

Reply:
This discussion makes me wonder what kind of collaborative project mathlib is. In the not exactly similar context of scientific publication, what you hint at (envisioning blocking something because a more general thing is in view) is considered as plainly unacceptable. The normal thing to do would be to mention the expected generalization and to ask whether it can be reached along this line.

Reply:
I completely agree that blocking your work here would be plainly unacceptable. I simply mean that Lason's version was incomplete by the time you finished Alon's version, and I needed to prioritise other things instead of continuing to work on that.

Reply:
If I understand correctly, there are three different ongoing works on this kind of ideas:
* the present PR by Bolton on Schwartz-Zippel
* my PR on the combinatorial Nullstellensatz
* subsequent work, eventually by Bhavik and Yaël, on a joint generalization.
May I suggest that then, things are done a bit differently? 
Is it worth the time and effort to commit neat PRs that will soon be superseded or should we join forces towards the envisioned general case? (Of course, that depends upon what “soon” means. And I'm wary of the potential dark holes created by good-will assertions such as “I have this and that good idea in mind that would…”, I've seen too much of them in my other math life.)

Reply:
I don't quite agree with the third ongoing work - the truth is that I have a 300 line file which exists solely on my laptop, most of which is developing theory unrelated to the Nullstellensatz, but contains a statement of Lason's version at the end. This one gives a slight generalisation of the main result of your PR, and is unrelated to the present PR: there is no known statement of a Nullstellensatz that implies both Alon's Nullstellensatz and the Schwartz-Zippel lemma. This file also hasn't been touched in the last month.

More importantly however, the theory that I began to develop is completely independent of the theory developed in your PR: everything except the last 80 lines in your Nullstellensatz.lean file would not be superseded (ie that's 1.5k lines that I don't plan to alter), and indeed the work there consists of things which ought to be in mathlib.

I'm very happy to join forces here, but where things stand currently I have comparatively little to contribute, so it would ultimately be me attempting to claim credit where it's undeserved.

Reply:
I agree with your understanding. I should however point out that
1. There is not *one* general case, but *two*. Schwartz-Zippel here is generalised by a different statement to the one generalising the combinatorial nullstellensatz.
2. In light of point 1, I duplicated little to no effort with your combinatorial nullstellensatz PR. Bhavik duplicated a bit before you opened the PR, but he stopped.
3. Most of your work would not be lost if we want to work on the generalisation. Everything about Gröbner bases and monomial division would still be useful. Only the very last part where you deduce the combinatorial nullstellensatz itself would change.
4. Point 1 might actually be a bit of a lie because Bhavik and I have heard that Sammy Luo managed to generalise. I don't know much more than that, though.

Reply:
Luo's recent work is here: https://arxiv.org/abs/2311.08873. It doesn't seem to mention Schwartz-Zippel at all, instead generalising two other polynomial methods along with CN. I think it's a distraction to this discussion, although the work is interesting.

Reply:
> Bhavik duplicated a bit before you opened the PR, but he stopped.

I don't believe I duplicated anything - there aren't even any identical statements in my file, let alone proofs!

Reply:
> It means bounding above the set of solutions in a product type. 

If I'm not mistaken, the statement would be, for a product $$A=\prod_{i\in I}A_i$$ of nonempty finite sets of a domain $$R$$ and a polynomial $$ f \in R[(T_i)] $$, than the number $$ \delta_A(f)$$ of zeroes of $$f$$ in $$A$$ satisfies $$ \delta_A(f) / |A| \leq \sup_d (\sum d_i/|A_i|)$$, the supremum being taken on the support of $$f$$. (IFor $$f\neq 0$$, one can even take $$d$$ to be the largest monomial for the lexicographic degree.)

Reply:
Antoine, I've already proved this in this PR 😉 (except for the "largest in lexicographic degree" part)

Reply:
The lexicographic refinement has to be left for future work, maybe you can just add a note in the file?